compiler/ifcc.g4

```diff
@@ -2,40 +2,46 @@ grammar ifcc;
 
 axiom: prog;
 
-prog: fn+;
-fn: TYPE VARNAME '(' argsDef? ')' '{' content '}';
-content: (init | affectation | ifElse | whileDo | returnValue) content?;
+prog: fn+ EOF;
+fn: TYPE VARNAME '(' argsDef? ')' '{' content? '}';
+content: (init | affectation ';' | ifElse | whileDo | returnValue ';' | fnCall ';') content?;
 value: CONST | VARNAME | CHAR;
-returnValue: 'return' expression ';';
+returnValue: 'return' expression;
 init: TYPE declaration; 
 declaration: dec (',' dec)* ';' ; 
 dec: VARNAME ('=' expression)? ;
-affectation: VARNAME '=' expression ';' # affectationExpr;
+affectation: VARNAME '=' expression # affectationExpr;
 expression:
-	expression MULTDIV expression # expressionMultDiv
+	expression MULTDIVMOD expression # expressionMultDivMod
 	| expression ADDSUB expression # expressionAddSub
-	| expression '&=' expression # expressionAnd
-	| expression '|=' expression # expressionOr
+	| expression ('&=' | '&&') expression # expressionAnd
+	| expression ('|=' | '||') expression # expressionOr
 	| expression '^=' expression # expressionXor
 	| expression '==' expression # expressionEqual
 	| expression '!=' expression # expressionNotEqual
 	| expression '>' expression # expressionGreater
 	| expression '<' expression # expressionLess
+	| expression '>=' expression # expressionGreaterEqual
+	| expression '<=' expression # expressionLessEqual
 	| '(' expression ')' # expressionPar
-	| VARNAME '(' args? ')' # expressionFn
+	| fnCall # expressionFn
 	| value # expressionValue;
-ifElse: 'if' '(' expression ')' '{' content '}' ( 'else' '{' content '}' )?;
+fnCall: VARNAME '(' args? ')';
+ifElse: 'if' '(' expression ')' '{' content? '}' ( 'else' '{' content? '}' )?;
 whileDo: 'while' '(' expression ')' '{' content '}';
 args: (expression) (',' expression)*;
 argsDef: (TYPE VARNAME) (',' TYPE VARNAME)*;
 
-CONST: [0-9]+;
-ADDSUB: '+' | '-';
-MULTDIV: '*' | '/';
-CHAR: '\'' .? '\'';
 COMMENT: '/*' .*? '*/' -> skip;
+ONELINECOMMENT: '//' .*? '\n' -> skip;
 DIRECTIVE: '#' .*? '\n' -> skip;
 WS: [ \t\r\n] -> channel(HIDDEN);
-ARITH: '+' | '-' | '*' | '/' | '%';
-TYPE: 'int' | 'char';
-VARNAME: [a-zA-Z_][a-zA-Z0-9_]*; 
\ No newline at end of file
+CONST: [0-9]+;
+ADDSUB: '+' | '-';
+MULTDIVMOD: '*' | '/' | '%';
+CHAR: '\'' .? '\'';
+REF: '&';
+TYPE: 'int' | 'char' | 'float' | 'double' | 'void';
+VARNAME: [a-zA-Z_]([a-zA-Z0-9_]| '\\u' HEX4 |'\\U' HEX4 HEX4)*;
+HEX: [0-9a-fA-F];
+HEX4: HEX HEX HEX HEX;
```

Comment:
Same here, I don't understand the point of these they do not seem to be used anywhere in the grammar or code

Reply:
That is the way used by gcc to allow special characters on the identificators (variable and functions names)